### ISSUE TITLE
Universal rendering under SSG cause miss-matched paths (nuxt/nuxt)

### ISSUE THREAD

**Issue by IO-Fire** [bloggrify/bloggrify#30]
See comment on issue outlined in nuxt/nuxt https://github.com/nuxt/nuxt/issues/15462#issuecomment-2526423630.

The benefits of fixing are considerable for SEO and page load performance.

### Example

- Every site page that is not the root
  - [https://minimalist.bloggrify.com/archives](https://minimalist.bloggrify.com/archives)
  - [https://minimalist.bloggrify.com/about](https://minimalist.bloggrify.com/about)
  - [https://minimalist.bloggrify.com/dummy](https://minimalist.bloggrify.com/dummy)
  - ...
  - [https://mistral.bloggrify.com/archives](https://mistral.bloggrify.com/archives)
  - [https://mistral.bloggrify.com/features](https://mistral.bloggrify.com/features)
  - ...
  - [https://epoxia.bloggrify.com/archives](https://epoxia.bloggrify.com/archives)
  - [https://epoxia.bloggrify.com/features](https://epoxia.bloggrify.com/features)
## Potential Fix

 **Potential SSG workaround solution with trailing slashes** in https://github.com/nuxt/nuxt/issues/15462#issuecomment-2526423630 is my preferred solution as it requires no side-channel implementations such as web server rewrites. 

> [!IMPORTANT]
>  Adopting trailing slashes requires all themes to adopt this approach. 

> [!WARNING]
> Exiting sites / themes directly using NuxtLink may require trailing slash enabled in `nuxt.config.ts`.

**Comment by IO-Fire** [comment]
I have been using **Potential SSG workaround solution with trailing slashes** from https://github.com/nuxt/nuxt/issues/15462#issuecomment-2526423630 in production and it is working well. 

Search engines started indexing more pages. 

On a non blog site, lighthouse scores increased by ~20 points on mobile.

**Comment by hlassiege** [comment]
Hi,
How did you encounter this problem ?

I mean :
- the sitemap should not include pages with trailing slash
- all pages (related to content) have a canonical link without trailing slash
-  the rss feed should not include pages with trailing slash

So, indeed, I can type manually the url with a trailing slash, but in a normal situation, you should never have those urls, no ?

For example, on https://eventuallycoding.com (my own blog), I don't see any duplicate content detected by google search console. 
Even if someone type the url with a trailing slash, it is automatically redirected to the page without trailing slash (I don't remember how it works, I may have enabled an option on netlify)

Isn't it ok ?

**Comment by IO-Fire** [comment]
Firstly, thank you for the reply. This issue is difficult to identify without explicitly looking for it.

> Hi, How did you encounter this problem ?

My own production websites have been affected by this issue. This alerted me to the problem that is present in Bloggrify in certain deployments.

> I mean :
> 
> * the sitemap should not include pages with trailing slash
> * all pages (related to content) have a canonical link without trailing slash
> * the rss feed should not include pages with trailing slash
> 
> So, indeed, I can type manually the url with a trailing slash, but in a normal situation, you should never have those urls, no ?

While the site doesn't reference the trailing slash paths, it is still accessed. 

The SSG page paths do not match the CSR paths. 

In the case of Bloggrify sites the pages are generated in two main ways CSR, SSR / SSG to from [universal rendering](https://nuxt.com/docs/guide/concepts/rendering#universal-rendering).

On the client with SSR / SSG, the web browser fetches the page from the server and renders it. CSR then enables. 

On the client with CSR, the app acts as an [SPA](https://en.wikipedia.org/wiki/Single-page_application), where the path is updated locally and the page contents is changed in the existing HTML document. 

Where the issue arrises:
When the pages are generated as `<path>/index.html`, a default web server configuration doesn't traditionally translate the path to page. 

When using a trailing slash the web server refers to the index file. 

When loading without a trailing slash the web server looks for a file or web server route (configured on the web server). 

## Worked example:

### SSR Side:

On the web server the path `/achives` is requested, the only result is a directory. The web server configuration states that a redirect is the action to take. It informs the client the location is redirected to `/archives/`.

### CSR Side:

The page is loaded. The path does not match app routes and the path is updated locally.

The redirect is difficult to notice as the CSR redirect is fast. 


## Results

Search engines rely on ["Signals" for Canonical URL Selection](https://www.searchenginejournal.com/google-uses-about-40-signals-to-determine-canonical-urls/534581/). 

For the Bloggrify demo sites, the canonical URL is being ignore in favour of the redirect destination URL. Visit [mistral bloggrify archives | Google Search](https://www.google.com/search?q=mistral+bloggrify+archives). Hover over the link to the archives page, It shows `https://mistral.bloggrify.com/archives/`.

The redirect destination is the page with the any content as the redirection page body is empty.

> For example, on https://eventuallycoding.com (my own blog), I don't see any duplicate content detected by google search console. Even if someone type the url with a trailing slash, it is automatically redirected to the page without trailing slash (I don't remember how it works, I may have enabled an option on netlify)

`eventuallycoding.com` is not affected by the first issue. The site is serving the requested non-trailing slash pages, something that [`mistral`](https://mistral.bloggrify.com/archives/), [`minimalist`](https://minimalist.bloggrify.com/archives/) are not. 

This may be due to mistral demo and minimalist demo not using the netlify nitro build preset it may include the `autoSubfolderIndex: false` discussed in https://github.com/nuxt/nuxt/issues/15462 also and available from [Deploy Nuxt to Cloudflare](https://nuxt.com/deploy/cloudflare). I haven't tested netlify nitro build preset.

While `eventuallycoding.com` is serving duplicated content it likely being ignored from indexing as `eventuallycoding.com` canonical URLs are the served pages. However the trailing slash pages may still be crawled if detected by search engines, using crawl budget though. Anecdotal evidence shows that Google Web Crawler is more advanced at finding the correct content address.

Some other sites are serving pages with and without trailing slashes, including [nuxt.com](https://nuxt.com/).
 
```bash
# Trailing slash
curl -I 'https://eventuallycoding.com/resources/'

# Shows code 200

# Non Trailing slash
curl -I 'https://eventuallycoding.com/resources'

# Shows code 200
```

> Isn't it ok ?

Yes. The duplicated content is likely ignored from indexing. Nitro presets may be updated in the future to prevent duplicate pages being served.

**Comment by IO-Fire** [comment]
I expect the solution will be autoSubfolderIndex from https://nuxt.com/deploy/cloudflare

**Comment by hlassiege** [comment]
Side note : 
On netlify, the problems comes from this setting apparently 

![Image](https://github.com/user-attachments/assets/99844e2d-6391-4959-b4a6-ddf72be68248)

Once you disable it, you don't have this redirect from /archives to /archives/

Before: 
```
curl -I 'https://mistral.bloggrify.com/archives'
HTTP/2 301 
```
After:
```
curl -I 'https://mistral.bloggrify.com/archives'
HTTP/2 200 
```

**Comment by hlassiege** [comment]
I've tried different things but I'm not completely satisfied

```
    experimental: {
        defaults: {
            nuxtLink: {
                trailingSlash: 'remove',
            }
        }
    },
```

This code should ensure that nuxtlink don't add trailing slash. But it's not really a solution for the whole problem.

I've added a middleware found on the issue listed above 

```js
export default defineNuxtRouteMiddleware((to, from) => {
    if (to.path !== '/' && to.path.endsWith('/')) {
        const { path, query, hash } = to
        const nextPath = path.replace(/\/+$/, '') || '/'
        const nextRoute = { path: nextPath, query, hash }
        return navigateTo(nextRoute, { redirectCode: 301 })
    }
})

```
It's nice when using SSR (or edge deployment), but it's useless with SSG.

**Comment by hlassiege** [comment]
Of course, the use of canonical urls is helping and I should not have duplicate content.
But having two working pages /archives and /archives/ still lead to wasted crawling budget.
It's not absolutely horrible, google bot knows how to deal with that. 
But still, it's not optimal.

I wonder if the solution is not on the hosting provider itself.
On a SSG site, I'm not sure we can solve this issue and add automatic redirect from archives/ to archives

Temporarily, I'll close this issue (and the associated pull request). To me, it's okkissh (not perfect, but working)
And expect Nuxt to come with a proper solution in the future
